asdf/commands/edit.py

```diff
@@ -0,0 +1,529 @@
+    # 4 bytes flag
+    # 4 bytes compression
+    # 8 bytes allocated size
+    # 8 bytes used (on disk) size
+    # 8 bytes data size
+    # 16 bytes checksum
+    bmlen = len(constants.BLOCK_MAGIC)
+    min_header = bmlen + 2 + 48  
+    uidx = 22
+    bindex = [start]
+    k = 0
+    while len(asdf_blocks) - k > min_header :
+        hsz = struct.unpack(">H",asdf_blocks[k+bmlen:k+bmlen+2])[0]
+        used = struct.unpack(">Q",asdf_blocks[k+uidx:k+uidx+8])[0]
+        k = k + bmlen + 2 + hsz + used
+        if k+bmlen > len(asdf_blocks) :
+            break
+        if constants.BLOCK_MAGIC==asdf_blocks[k:k+bmlen] :
+            bindex.append(k+start)
+        else :
+            break
+    return bindex, asdf_blocks[:k]
+
+    
+def write_block_index ( fd, index ) :
+    if len(index) < 1 :
+        return
+
+    bindex_hdr = b'#ASDF BLOCK INDEX\n%YAML 1.1\n---\n'
+    fd.write(bindex_hdr)
+    for idx in index :
+        ostr = f'- {idx}\n'
+        fd.write(ostr.encode('utf-8'))
+    end = b'...'
+    fd.write(end)
+    return
+
+
+def rewrite_asdf_file ( edited_text, orig_text, oname, fname ) :
+    """ Rewrite an ASDF file for too large edited YAML.  The edited YAML, a buffer,
+    the blocks will be rewritten.  A block index will also be rewritten.  If a
+    block index existed in the old file, it will have to be recomputed to 
+    because of the larger YAML size and buffer, which changes the location of 
+    the binary blocks.
+
+    Parameters
+    ----------
+    edited_text : the new YAML text to write out.
+    orig_text : the original YAML text to overwrite.
+    oname : the ASDF file to overwrite.
+    fname : the edit YAML to write to new file.
+    """
+    
+    tmp_oname = oname + '.tmp' # Save as a temp file, in case anything goes wrong.
+    buffer_size = 10 * 1000
+    buffered_text = add_buffer_to_new_text(edited_text,buffer_size) 
+
+    with open(oname,"r+b") as fd :
+        orig_buffer = fd.read() # Small enough to simply read the  whole thing
+    
+    # Get the binary blocks, compute the new block index, and strip old block
+    # index, if it exists.
+    asdf_blocks = orig_buffer[len(orig_text):] 
+    index, asdf_blocks = compute_block_index_blocks(len(buffered_text),asdf_blocks)
+    out_bytes = buffered_text + asdf_blocks
+
+    # Write new file with edited text, buffer, and recomputed block index.
+    with open(tmp_oname,"w+b") as fd :
+        fd.write(out_bytes)
+        write_block_index(fd,index)
+
+    # Rename temp file.
+    os.rename(tmp_oname,oname)
+
+    # Output message to user.
+    delim = '*' * 70
+    print(f"\n{delim}")
+    print(f"The text in '{fname}' was too large to simply overwrite the")
+    print(f"text in '{oname}'.  The file '{oname}' was rewritten to")
+    print(f"accommodate the larger text size.  Also, {buffer_size:,} bytes")
+    print(f"as a buffer for the text in '{oname}' to allow for future edits.")
+    print(f"{delim}\n")
+
+def save_func ( fname, oname ) :
```

Comment:
I'm not sure the large file bailout makes any sense. Someone has just spent time editing a header and then is told, sorry I can't use it. I'm going to guess that in almost all circumstances they have spent more time editing then it will take to rewrite the file. If there isn't enough disk space left, that's one issue, but otherwise, just do it (IMHO)

Reply:
I agree with this.  Since available memory may well be an issue, we should copy the bytes from old to new in batches.

Reply:
I changed this to rewrite in this case.